cloudimg/ms_azure.py

```diff
@@ -0,0 +1,513 @@
+
+            if not filtered_blob:
+                log.error("Image not found with tags \"%s\"", metadata.tags)
+                blob = self.upload_to_container(
+                    image_path=metadata.image_path,
+                    container_name=metadata.container,
+                    object_name=metadata.object_name,
+                    tags=metadata.tags,
+                )
+            else:
+                log.info(
+                    "Image already exists on container \"%s\""
+                    " with name \"%s\"",
+                    filtered_blob.container_name,
+                    filtered_blob.name,
+                )
+
+                blob = self.get_object_by_name(
+                    container=filtered_blob.container_name,
+                    name=filtered_blob.name,
+                )
+
+        else:
+            log.info("Image already exists with name \"%s\"", blob.blob_name)
+
+        log.info('Image published: %s/%s', blob.container_name, blob.blob_name)
+        return blob.get_blob_properties()
+
+    def get_blob_sas_uri(self, blob_props):
+        """
+        Get a SAS URL of a blob using its Storage account's connection string
+        The SAS URL is with read-only permissions that will expire
+        in `self.SAS_EXPIRE` year(s).
+
+        Args:
+            container_name (str): The blob's container name
+            blob_props (BlobProperties): The uploaded blob properties
+        Returns:
+            The blob's full SAS URI
+        """
+        logging.info(
+            "Generating the SAS URI for %s/%s",
+            blob_props.container,
+            blob_props.name,
+        )
+        uri = generate_blob_sas(
+            account_name=self._account_name,
+            account_key=self._account_key,
+            container_name=blob_props.container,
+            blob_name=blob_props.name,
+            permission=BlobSasPermissions(read=True),
+            expiry=datetime.utcnow() + timedelta(days=self.SAS_EXPIRY),
+        )
+        # Full SAS format:
+        # https://learn.microsoft.com/en-us/azure/storage/common/storage-sas-overview#sas-token
+        full_sas_uri = (
+            "https://%s.blob.core.windows.net/%s/%s?%s" % (
+                self._account_name,
+                blob_props.container,
+                blob_props.name,
+                uri,
+            )
+        )
+        return full_sas_uri
+
+    def tag_image(self, blob_props, tags):
+        """
+        Apply the corresponding tags to an uploaded blob
+        Args:
+            blob_props (BlobProperties): The uploaded blob properties
+            tags (dict): Dictionary with the tags to apply
+        Returns:
+           Dictionary with updated properties
+        """
+        blob = self.get_object_by_name(
+                    container=blob_props.container,
+                    name=blob_props.name
+               )
+
+        if not blob:
```

Comment:
Not treating this as an error seems rather surprising.  Are you sure you want to allow this?

If you're sure, maybe the warning should at least include the container and blob name?  And the doc string should probably say that the blob is updated "if it exists".

Reply:
You're right!

I've changed it to raise an exception when the blob is not found.